File: Easy/141. Linked List Cycle/141. Linked List Cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
    if (!head || !head.next) {
        return false; // No cycle in an empty list or a list with a single node.
    }

    let slow = head; // The slow pointer moves one step at a time.
    let fast = head; // The fast pointer moves two steps at a time.

    while (fast && fast.next) {
        slow = slow.next; // Move slow by one step.
        fast = fast.next.next; // Move fast by two steps.

        if (slow === fast) {
            return true; // If there is a cycle, slow and fast will meet.
        }
    }

    return false; // If fast reaches the end, there is no cycle.
};

File: Easy/141. Linked List Cycle/141. Linked List Cycle.rs
// Definition for singly-linked list.
#[derive(Debug, PartialEq, Eq)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn has_cycle(head: Option<Box<ListNode>>) -> bool {
        let (mut slow, mut fast) = (head.as_ref(), head.as_ref());

        while let (Some(node1), Some(node2)) = (slow, fast) {
            slow = node1.next.as_ref();
            fast = node2.next.as_ref().and_then(|n| n.next.as_ref());

            if slow == fast {
                return true; // If there is a cycle, slow and fast will meet.
            }
        }

        false // If fast reaches the end, there is no cycle.
    }
}

// [warn] rust section is not available for this problem yet so this code is not tested.
